File: onyx-database-tests/test/embedded/SkipListTest.java
package embedded;

import com.onyx.structure.DefaultMapBuilder;
import com.onyx.structure.MapBuilder;
import com.onyx.structure.store.StoreType;
import org.junit.FixMethodOrder;
import org.junit.Ignore;
import org.junit.Test;
import org.junit.runners.MethodSorters;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.BiConsumer;

/**
 * Created by tosborn1 on 1/6/17.
 */
@FixMethodOrder(MethodSorters.NAME_ASCENDING)
public class SkipListTest
{

    public static final String TEST_DATABASE = "C:/Sandbox/Onyx/Tests/skip.db";

    @Test
    public void testInsert()
    {
        MapBuilder builder = new DefaultMapBuilder(TEST_DATABASE, StoreType.MEMORY_MAPPED_FILE);
        Map<Integer, Integer> skipList = builder.getSkipListMap("first");

        Map keyValues = new HashMap();
        for(int i = 0; i < 100000; i++)
        {
            int randomNum = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE);
            int randomValue = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE);

            skipList.put(randomNum, randomValue);
            keyValues.put(randomNum, randomValue);
        }
        keyValues.forEach((o, o2) -> {
            assert skipList.get((Integer)o).equals(o2);
        });
    }

    @Test
    public void testDelete()
    {


        MapBuilder builder = new DefaultMapBuilder(TEST_DATABASE);
        Map<Integer, Integer> skipList = builder.getSkipListMap("second");

        Map keyValues = new HashMap();
        for(int i = 0; i < 50000; i++)
        {
            int randomNum = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE);
            int randomValue = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE);

            skipList.put(randomNum, randomValue);
            keyValues.put(randomNum, randomValue);

        }
        Map newKeyValues = new HashMap();
        Map deletedKeyValues = new HashMap();

        final AtomicInteger val = new AtomicInteger(0);
        keyValues.forEach((o, o2) -> {

            assert skipList.get((Integer)o).equals(o2);

            if((val.addAndGet(1) % 1000) == 0)
            {
                skipList.remove((Integer)o);
                deletedKeyValues.put(o, o2);
            }
            else
            {
                newKeyValues.put(o, o2);
            }
        });

        newKeyValues.forEach((o, o2) -> {
            assert skipList.get((Integer)o).equals(o2);
        });

        deletedKeyValues.forEach((o, o2) -> {
            assert skipList.get((Integer)o) == null;
        });

    }

    @Test
    public void testUpdate()
    {
        MapBuilder builder = new DefaultMapBuilder(TEST_DATABASE);
        Map<Integer, Integer> skipList = builder.getSkipListMap("third");

        Map keyValues = new HashMap();
        for(int i = 0; i < 10000; i++)
        {
            int randomNum = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE);
            int randomValue = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE);

            skipList.put(randomNum, randomValue);
            keyValues.put(randomNum, randomValue);

        }

        keyValues.forEach((o, o2) -> skipList.put((Integer)o, (Integer)o2));
    }
    @Test
    public void testForEach()
    {
        MapBuilder builder = new DefaultMapBuilder(TEST_DATABASE);
        Map<Integer, Integer> skipList = builder.getSkipListMap("third");

        Map keyValues = new HashMap();
        for(int i = 0; i < 50000; i++)
        {
            int randomNum = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE);
            int randomValue = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE);

            skipList.put(randomNum, randomValue);
            keyValues.put(randomNum, randomValue);

        }

        final AtomicInteger val = new AtomicInteger(0);
        keyValues.forEach((o, o2) -> {

            assert skipList.get((Integer)o).equals(o2);

            if((val.addAndGet(1) % 1000) == 0)
            {
                skipList.remove((Integer)o);
            }
        });


        final AtomicInteger numberOfValues = new AtomicInteger(0);
        skipList.forEach((integer, integer2) -> {
            assert integer != null;
            assert integer2 != null;
            numberOfValues.addAndGet(1);
        });


        assert numberOfValues.get() == skipList.size();
    }

    @Test
    public void testKeyIterator()
    {
        MapBuilder builder = new DefaultMapBuilder(TEST_DATABASE);
        Map<Integer, Integer> skipList = builder.getSkipListMap("fourth");

        Map keyValues = new HashMap();
        for(int i = 0; i < 50000; i++)
        {
            int randomNum = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE);
            int randomValue = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE);

            skipList.put(randomNum, randomValue);
            keyValues.put(randomNum, randomValue);
        }

        final AtomicInteger val = new AtomicInteger(0);
        keyValues.forEach((o, o2) -> {

            assert skipList.get((Integer)o).equals(o2);

            if((val.addAndGet(1) % 1000) == 0)
            {
                skipList.remove((Integer)o);
            }
        });


        final AtomicInteger numberOfValues = new AtomicInteger(0);

        Iterator<Integer> iterator = skipList.keySet().iterator();
        while(iterator.hasNext())
        {
            assert iterator.next() instanceof Integer;
            numberOfValues.addAndGet(1);
        }


        assert numberOfValues.get() == skipList.size();
    }

    @Test
    public void testValueIterator()
    {
        MapBuilder builder = new DefaultMapBuilder(TEST_DATABASE);
        Map<Integer, Integer> skipList = builder.getSkipListMap("fifth");

        Map keyValues = new HashMap();
        for(int i = 0; i < 50000; i++)
        {
            int randomNum = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE);
            int randomValue = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE);

            skipList.put(randomNum, randomValue);
            keyValues.put(randomNum, randomValue);

        }

        final AtomicInteger val = new AtomicInteger(0);
        keyValues.forEach((o, o2) -> {

            assert skipList.get((Integer)o).equals(o2);

            if((val.addAndGet(1) % 1000) == 0)
            {
                skipList.remove((Integer)o);
            }
        });


        final AtomicInteger numberOfValues = new AtomicInteger(0);

        Iterator<Integer> iterator = skipList.values().iterator();
        while(iterator.hasNext())
        {
            assert iterator.next() instanceof Integer;
            numberOfValues.addAndGet(1);
        }


        assert numberOfValues.get() == skipList.size();
    }
}

File: onyx-database-tests/test/embedded/SkipListTest.kt
package embedded

import com.onyx.exception.InitializationException
import com.onyx.diskmap.factory.impl.DefaultDiskMapFactory
import com.onyx.diskmap.factory.DiskMapFactory
import org.junit.FixMethodOrder
import org.junit.Test
import org.junit.runners.MethodSorters

import java.util.HashMap
import java.util.concurrent.ThreadLocalRandom
import java.util.concurrent.atomic.AtomicInteger

/**
 * Created by Tim Osborn on 1/6/17.
 */
@FixMethodOrder(MethodSorters.NAME_ASCENDING)
class SkipListTest {

    @Test
    @Throws(InitializationException::class)
    fun testInsert() {

        val builder = DefaultDiskMapFactory(TEST_DATABASE)
        val skipList = builder.getSkipListMap<MutableMap<Int, Int>>("first")

        val keyValues = HashMap<Int, Int>()
        for (i in 0..49999) {
            val randomNum = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE)
            val randomValue = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE)

            skipList.put(randomNum, randomValue)
            keyValues.put(randomNum, randomValue)
        }


        val time = System.currentTimeMillis()
        keyValues.forEach({ o, o2 -> assert(skipList[o] == o2) })
    }

    @Test
    fun testDelete() {


        val builder = DefaultDiskMapFactory(TEST_DATABASE)
        val skipList = builder.getSkipListMap<MutableMap<Int, Int>>("second")

        val keyValues = HashMap<Int, Int>()
        for (i in 0..49999) {
            val randomNum = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE)
            val randomValue = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE)

            skipList.put(randomNum, randomValue)
            keyValues.put(randomNum, randomValue)

        }
        val newKeyValues = HashMap<Int, Int>()
        val deletedKeyValues = HashMap<Int, Int>()

        val `val` = AtomicInteger(0)
        keyValues.forEach({ o, o2 ->

            assert(skipList[o] == o2)

            if (`val`.addAndGet(1) % 1000 == 0) {
                skipList.remove(o)
                deletedKeyValues.put(o, o2)
            } else {
                newKeyValues.put(o, o2)
            }
        })

        newKeyValues.forEach({ o, o2 -> assert(skipList[o] == o2) })

        deletedKeyValues.forEach({ o, o2 -> assert(skipList[o] == null) })

    }

    @Test
    fun testUpdate() {
        val builder = DefaultDiskMapFactory(TEST_DATABASE)
        val skipList = builder.getSkipListMap<MutableMap<Int, Int>>("third")

        val keyValues = HashMap<Int, Int>()
        for (i in 0..9999) {
            val randomNum = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE)
            val randomValue = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE)

            skipList.put(randomNum, randomValue)
            keyValues.put(randomNum, randomValue)

        }

        keyValues.forEach({ o, o2 -> skipList.put(o as Int, o2 as Int) })
    }

    @Test
    fun testForEach() {
        val builder = DefaultDiskMapFactory(TEST_DATABASE)
        val skipList = builder.getSkipListMap<MutableMap<Int, Int>>("third")

        val keyValues = HashMap<Int, Int>()
        for (i in 0..49999) {
            val randomNum = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE)
            val randomValue = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE)

            skipList.put(randomNum, randomValue)
            keyValues.put(randomNum, randomValue)

        }

        val `val` = AtomicInteger(0)
        keyValues.forEach({ o, o2 ->

            assert(skipList[o] == o2)

            if (`val`.addAndGet(1) % 1000 == 0) {
                skipList.remove(o)
            }
        })


        val numberOfValues = AtomicInteger(0)
        skipList.forEach { integer, integer2 ->
            assert(integer != null)
            assert(integer2 != null)
            numberOfValues.addAndGet(1)
        }


        assert(numberOfValues.get() == skipList.size)
    }

    @Test
    fun testKeyIterator() {
        val builder = DefaultDiskMapFactory(TEST_DATABASE)
        val skipList = builder.getSkipListMap<MutableMap<Int, Int>>("fourth")

        val keyValues = HashMap<Int, Int>()
        for (i in 0..49999) {
            val randomNum = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE)
            val randomValue = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE)

            skipList.put(randomNum, randomValue)
            keyValues.put(randomNum, randomValue)
        }

        val `val` = AtomicInteger(0)
        keyValues.forEach({ o, o2 ->

            assert(skipList[o] == o2)

            if (`val`.addAndGet(1) % 1000 == 0) {
                skipList.remove(o)
            }
        })


        val numberOfValues = AtomicInteger(0)

        val iterator = skipList.keys.iterator()
        while (iterator.hasNext()) {
            assert(iterator.next() is Int)
            numberOfValues.addAndGet(1)
        }


        assert(numberOfValues.get() == skipList.size)
    }

    @Test
    fun testValueIterator() {
        val builder = DefaultDiskMapFactory(TEST_DATABASE)
        val skipList = builder.getSkipListMap<MutableMap<Int, Int>>("fifth")

        val keyValues = HashMap<Int, Int>()
        for (i in 0..49999) {
            val randomNum = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE)
            val randomValue = ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE)

            skipList.put(randomNum, randomValue)
            keyValues.put(randomNum, randomValue)

        }

        val `val` = AtomicInteger(0)
        keyValues.forEach({ o, o2 ->

            assert(skipList[o] == o2)

            if (`val`.addAndGet(1) % 1000 == 0) {
                skipList.remove(o)
            }
        })


        val numberOfValues = AtomicInteger(0)

        val iterator = skipList.values.iterator()
        while (iterator.hasNext()) {
            assert(iterator.next() is Int)
            numberOfValues.addAndGet(1)
        }


        assert(numberOfValues.get() == skipList.size)
    }

    companion object {

        val TEST_DATABASE = "C:/Sandbox/Onyx/Tests/skip.db"
    }
}
